libtenzir/include/tenzir/detail/adaptive_table_slice_builder_guards.hpp

```diff
@@ -129,26 +133,34 @@ class list_guard {
   template <concrete_type Type>
   auto add_impl(auto view) -> caf::error {
     if (not value_type)
-      propagate_type(tenzir::type{Type{}});
+      propagate_type_discovery(tenzir::type{Type{}});
     return caf::visit(
       detail::overload{
         [this, view](const Type& t) {
           append_value_to_builder(t, view);
           return caf::error{};
         },
         [view, this](const auto& t) {
-          if (auto castable = can_cast(Type{}, t); not castable)
-            return std::move(castable.error());
           auto cast_val = cast_value(Type{}, view, t);
-          if (not cast_val)
-            return std::move(cast_val.error());
-          // this is sometimes not used due to the if constexpr below.
-          // This is done to prevent compilation warning treated as an error.
-          (void)this;
-          if constexpr (not std::is_same_v<caf::expected<void>,
-                                           decltype(cast_val)>) {
-            append_value_to_builder(t, *cast_val);
+          if (cast_val) {
+            // "this" is sometimes not used due to the if constexpr below.
+            // It prevents compilation warning treated as an error.
+            (void)this;
+            // We don't handle caf::expected<void> here because the cast api
+            // only returns caf::expected<void> if the operation is not
+            // supported. The caf::expected<void> will fall into common_type
+            // casting case.
```

Comment:
This seems brittle to me from an API user perspective. I would expect a cast to type T to return me an `expected<type_to_data_t<T>>`, regardless of whether that cast is possible or not.

Reply:
done in d19a6661b1904fd4398fd4caac06f2caa10c5bd3